scrapy/settings/__init__.py

```diff
@@ -174,11 +251,28 @@ def __iter__(self, k, v):
         return iter(self.o)
 
 
+class Settings(BaseSettings):
+
+    def __init__(self, values=None, priority='project'):
+        # Do not pass kwarg values here. We don't want to promote user-defined
+        # dicts, and we want to update, not replace, default dicts with the
+        # values given by the user
+        super(Settings, self).__init__()
+        self.setmodule(default_settings, 'default')
+        # Promote default dictionaries to BaseSettings instances for per-key
+        # priorities
+        for name in self.attributes:
+            val = self[name]
+            if isinstance(val, dict):
+                self.set(name, BaseSettings(val, 'default'), 'default')
+        self.setdict(values, priority)
+
+
 class CrawlerSettings(Settings):
 
     def __init__(self, settings_module=None, **kw):
-        Settings.__init__(self, **kw)
         self.settings_module = settings_module
+        Settings.__init__(self, **kw)
```

Comment:
Yes, took me quite a while to figure out though ;D It became necessary after I moved the dict promotion into `__init__()`. `Settings.__init__()` uses `__getitem__()` during the promotion of default dictionaries to `BaseSettings` instances, which in turn accesses `self.settings_module`, so it needs to be defined before calling `__init__()`